Frameworks/Foundation/NSUndoManager.mm

```diff
@@ -38,83 +104,231 @@ - (id)prepareWithInvocationTarget:(id)target {
+        [_undoStack popBack];
+    }
+    if (_groupsByEvent && !_isAutomaticGroupOpen) {
+        [self _createAutomaticUndoGroup];
+    }
+    _NSUndoGroup* undoGroup = [_undoStack peek];
+
+    if (undoGroup == nil || [undoGroup isClosed]) {
+        _NSUndoGroup* newGroup = [[_NSUndoGroup alloc] init];
+        [_undoStack push:newGroup];
+        [newGroup release];
+        _undoGroupingLevel = 1;
+        _hasNestedGroup = NO;
+    } else {
+        [undoGroup createUndoGroup];
+        _undoGroupingLevel++;
+        _hasNestedGroup = YES;
+    }
+    if (!_redoing) {
+        _groupingLevel = _undoGroupingLevel;
+    }
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)endUndoGrouping {
-    UNIMPLEMENTED();
+    if (_undoing) {
+        return;
+    }
+    // No group to close.
+    if (_undoGroupingLevel == 0) {
+        [NSException raise:NSInternalInconsistencyException format:@"endUndoGrouping called with no open group."];
+    }
+    _NSUndoGroup* undoGroup = [_undoStack peek];
+    [undoGroup closeUndoGroup];
+    _undoGroupingLevel--;
+    if (!_redoing) {
+        _groupingLevel = _undoGroupingLevel;
+    }
+}
+
+// Internal redo grouping
+- (void)_beginRedoGrouping {
+    _NSUndoGroup* redoGroup = [_redoStack peek];
+
+    if (redoGroup == nil || [redoGroup isClosed]) {
+        _NSUndoGroup* newGroup = [[_NSUndoGroup alloc] init];
+        [_redoStack push:newGroup];
+        [newGroup release];
+    } else {
+        [redoGroup createUndoGroup];
+    }
+    _redoGroupingLevel++;
+    if (!_redoing) {
+        _groupingLevel = _redoGroupingLevel;
+    }
+}
+
+- (void)_endRedoGrouping {
+    // No group to close.
+    if (_redoGroupingLevel == 0) {
+        [NSException raise:NSInternalInconsistencyException format:@"Attempted to close a redo group with no open redo group."];
+    }
+    _NSUndoGroup* redoGroup = [_redoStack peek];
+    [redoGroup closeUndoGroup];
+    _redoGroupingLevel--;
+    if (!_redoing) {
+        _groupingLevel = _redoGroupingLevel;
+    }
 }
 
 /**
- @Status Stub
- @Notes
+@Status Interoperable
+*/
+- (BOOL)isUndoRegistrationEnabled {
+    @synchronized(self) {
+        return _disableCount == 0;
+    }
+}
+
+/**
+ @Status Interoperable
 */
 - (void)disableUndoRegistration {
-    UNIMPLEMENTED();
+    @synchronized(self) {
+        _disableCount++;
+    }
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)enableUndoRegistration {
-    UNIMPLEMENTED();
+    @synchronized(self) {
+        if (_disableCount == 0) {
+            [NSException raise:NSInternalInconsistencyException
+                        format:@"Called enable undo registration while undo registration is enabled."];
+        }
+        _disableCount--;
+    }
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 - (void)removeAllActions {
-    UNIMPLEMENTED();
+    [self _removeUndoActions];
+    [self _removeRedoActions];
+}
+
+- (void)_removeUndoActions {
```

Comment:
Why is it `removeAllActions`' responsibility to set the grouping level to 0, but there's a special function called `_removeUndoActions` that doesn't?

It seems like you'd either:
1) have one function that clears all states
or
2) have one function per stack that resets _stack-specific_ state.

Right now, you have two functions, one that clears half the states and then individual ones that clear the other halves of the states.

---
In reply to: [76708543](https://github.com/Microsoft/WinObjC/pull/900#discussion_r76708543) [](ancestors = 76708543)